src/LinuxCaptureService/LinuxCaptureService.cpp

```diff
@@ -19,22 +19,32 @@ grpc::Status LinuxCaptureService::Capture(
 
   orbit_producer_event_processor::GrpcClientCaptureEventCollector
       grpc_client_capture_event_collector{reader_writer};
-
-  std::shared_ptr<orbit_capture_service_base::StartStopCaptureRequestWaiter>
-      start_stop_capture_request_waiter =
-          orbit_capture_service_base::CreateGrpcStartStopCaptureRequestWaiter(reader_writer);
-
   CaptureServiceBase::CaptureInitializationResult initialization_result =
-      DoCapture(&grpc_client_capture_event_collector, start_stop_capture_request_waiter);
+      InitializeCapture(&grpc_client_capture_event_collector);
   switch (initialization_result) {
     case CaptureInitializationResult::kSuccess:
-      return grpc::Status::OK;
+      break;
     case CaptureInitializationResult::kAlreadyInProgress:
       return {grpc::StatusCode::ALREADY_EXISTS,
               "Cannot start capture because another capture is already in progress"};
   }
 
-  ORBIT_UNREACHABLE();
+  std::shared_ptr<orbit_capture_service_base::StartStopCaptureRequestWaiter>
+      start_stop_capture_request_waiter =
+          orbit_capture_service_base::CreateGrpcStartStopCaptureRequestWaiter(reader_writer);
+  auto capture_options_or_error = start_stop_capture_request_waiter->WaitForStartCaptureRequest();
```

Comment:
Maybe I'm just missing some context on the cloud collector but something is strange here.
The point of `StartStopCaptureRequestWaiter` was that it's passed to `DoCapture`, method shared by `LinuxCaptureService` and the cloud collector. But now `WaitForStartCaptureRequest` is called outside of it, and `DoCapture` only calls `WaitForStopCaptureRequest`.
So is it even necessary now to have `WaitForStartCaptureRequest` in the interface? Doesn't the shared code between `LinuxCaptureService` and the cloud collector only use `WaitForStopCaptureRequest`? Now where `WaitForStartCaptureRequest` is used, the concrete type is known.

Reply:
The `CloudCollector::StartCaptureService` method (this can be found in code search) will implement the similar logic as the `LinuxCaptureService::Capture` and `WindowsCaptureService::Capture` methods. And hence we still use `WaitForStartCaptureRequest` in the CloudCollector.

Reply:
You still need the logic implemented in `WaitForStartCaptureRequest` by the implementations of `StartStopCaptureRequestWaiter`. What I'm saying is that you don't need that method to be part of the interface anymore, because it's not used in `DoCapture` (which doesn't know about the concrete types), but it's directly used in `LinuxCaptureService::Capture`, `WindowsCaptureService::Capture` and `CloudCollector::StartCaptureService`. All these methods use `WaitForStartCaptureRequest` already knowing the type of the `StartStopCaptureRequestWaiter`. For example, the content of `GrpcStartStopCaptureRequestWaiter::WaitForStartCaptureRequest` can be directly moved here. So the interface can become `StopCaptureRequestWaiter` and contain only `WaitForStopCaptureRequest` as that's the only method used under `DoCapture`.

Reply:
Got it! Thanks for the explaination!

Reply:
I removed `WaitForStartCaptureRequest` method from the interface. The suggestion is really great and now we can have different return type for `WaitForStartCaptureRequest`. Thanks again!